com.microsoft.java.lsif.core/src/com/microsoft/java/lsif/core/internal/visitors/LsifVisitor.java

```diff
@@ -71,19 +161,119 @@ private void resolve(int startPosition, int length, boolean needResolveImpl) {
+			ICompilationUnit compilationUnit = (ICompilationUnit) element.getAncestor(IJavaElement.COMPILATION_UNIT);
+			IClassFile cf = (IClassFile) element.getAncestor(IJavaElement.CLASS_FILE);
+			PackageManager manager = null;
+			String groupId = "";
+			String artifactId = "";
+			String version = "";
+			String type = "";
+			String url = "";
+			if (compilationUnit == null && cf != null) {
+				IPath path = cf.getPath();
+				IPackageFragmentRoot root = javaproject.findPackageFragmentRoot(path);
+				IClasspathEntry container = root.getRawClasspathEntry();
+				IPath containerPath = container.getPath();
+				String pathName = containerPath.toString();
+				if (pathName.startsWith("org.eclipse.jdt.launching.JRE_CONTAINER")) {
+					// JDK Library
+					manager = PackageManager.JDK;
+					Manifest manifest = new Manifest();
+					if (root instanceof JrtPackageFragmentRoot) {
+						// For JDK 9+
+						manifest = ((JrtPackageFragmentRoot) root).getManifest();
+						IModuleDescription moduleDescription = ((JrtPackageFragmentRoot) root).getAutomaticModuleDescription();
+						groupId = moduleDescription.getElementName();
+					} else if (root instanceof JarPackageFragmentRoot) {
+						// For JDK 8-
+						manifest = ((JarPackageFragmentRoot) root).getManifest();
+						IModuleDescription moduleDescription = ((JarPackageFragmentRoot) root).getAutomaticModuleDescription();
+						groupId = moduleDescription.getElementName();
+					}
+					if (manifest != null) {
+						Attributes attributes = manifest.getMainAttributes();
+						version = attributes.getValue("Implementation-Version");
+					}
+				} else {
+					List<File> fileList = new ArrayList<>();
+					if (builder == ProjectBuildTool.MAVEN) {
+						findPomFile(path.removeLastSegments(1).toFile(), fileList);
+					} else if (builder == ProjectBuildTool.GRADLE) {
+						findPomFile(path.removeLastSegments(2).toFile(), fileList);
+					}
+					if (fileList.size() > 0) {
+						manager = (builder == ProjectBuildTool.MAVEN) ? PackageManager.MAVEN : PackageManager.GRADLE;
+						File file = fileList.get(0);
+						MavenXpp3Reader reader = new MavenXpp3Reader();
+						Model model = reader.read(new FileReader(file.getAbsolutePath()));
+						groupId = model.getGroupId();
+						artifactId = model.getArtifactId();
+						version = model.getVersion();
+						Parent parent = model.getParent();
+						if (parent != null) {
+							if (groupId == null) {
+								groupId = parent.getGroupId();
+							}
+							if (artifactId == null) {
+								artifactId = parent.getArtifactId();
+							}
+							if (version == null) {
+								version = parent.getVersion();
+							}
+						}
+						Scm scm = model.getScm();
+						if (scm != null) {
+							url = scm.getUrl();
+							type = ScmUrlUtils.getProvider(scm.getConnection());
+						}
+					}
+				}
+			}
+			String schemeId = "";
+			if (manager == PackageManager.MAVEN || manager == PackageManager.GRADLE) {
+				schemeId = groupId + "/" + artifactId;
```

Comment:
I think we can combine `groupId ` and `artifactId` into one variable. Something like `packageName`